### ISSUE TITLE
Title: Query on EfficientSAM's Adaptability for Low-Resource Environments

### ISSUE THREAD

**Issue by yihong1120** [yformer/EfficientSAM#16]
Dear EfficientSAM Contributors,

I hope this message finds you well. I have been closely following the development of EfficientSAM and am thoroughly impressed with the strides made in efficient image segmentation. The recent release of the torchscript version and the accompanying Colab notebook have been particularly helpful.

However, I am curious about the model's performance and adaptability in low-resource environments, which is often a challenge in the field of computer vision. Specifically, I am interested in understanding how EfficientSAM fares on devices with limited computational power and memory constraints.

Could you provide insights or benchmarks on the following aspects?

1. The model's performance on CPUs with lower clock speeds and fewer cores, as compared to the recommended specifications.
2. The memory footprint of EfficientSAM when deployed in a constrained environment, and any potential trade-offs that might need to be considered.
3. Any recommended practices or modifications for optimising EfficientSAM's deployment on such devices without significantly compromising on accuracy or speed.

I believe addressing these queries would greatly benefit researchers and practitioners working in regions with limited access to high-end hardware, thereby broadening the scope of EfficientSAM's applicability.

Thank you for your time and the remarkable work on this project. I look forward to your response and any guidance you can provide.

Best regards,
yihong1120

**Comment by yformer** [comment]
@yihong1120, thanks for your interest in EfficientSAM! 

Can you give some detailed information on the hardware platform for your use case, e.g, memory constraints, latency constraints, and CPU type?

**Comment by WanBenLe** [comment]
When I download to model and export to A40 GPU, seems to use 35Gb GPU memory and sam without prompt, use over than 20s.
Maybe it's too slow to use.

**Comment by yformer** [comment]
@WanBenLe, can you kill all the jobs you are running to check?

**Comment by WanBenLe** [comment]
I used it:
efficient_sam_vits_model = build_efficient_sam_vits()
efficient_sam_vits_model.to('cuda:3')
 
In def predict_masks, take_along_dim will lead to out of GPU memory.
iou_predictions = torch.take_along_dim(iou_predictions, sorted_ids, dim=2)

So I have to 
output_masks=output_masks.to('cpu')
iou_predictions=iou_predictions.to('cpu')

I also use cuda to run in batched_nms process_small_region.
masks = torch.cat(new_masks, dim=0).to('cuda:3')
boxes = batched_mask_to_box(masks)
keep_by_nms = batched_nms(
    boxes.float(),
    torch.as_tensor(scores).to('cuda:3'),
    torch.zeros_like(boxes[:, 0]).to('cuda:3'),  # categories
    iou_threshold=nms_thresh,
)

33gb GPU Mem of model.
41gb GPU Mem of model +  batched_nms.

Maybe I had something wrong?

**Comment by yformer** [comment]
@WanBenLe What is the input for your case?

**Comment by WanBenLe** [comment]
Image, which resize 512*512

with zipfile.ZipFile("weights/efficient_sam_vits.pt.zip", 'r') as zip_ref:
    zip_ref.extractall("weights")
efficient_sam_vits_model = build_efficient_sam_vits()
efficient_sam_vits_model.to('cuda:2')
efficient_sam_vits_model.device
rle = run_everything_ours(cv2.resize(image,(512,512)), efficient_sam_vits_model)

**Comment by WanBenLe** [comment]
@yformer 
Image, which resize 512*512

with zipfile.ZipFile("weights/efficient_sam_vits.pt.zip", 'r') as zip_ref:
zip_ref.extractall("weights")
efficient_sam_vits_model = build_efficient_sam_vits()
efficient_sam_vits_model.to('cuda:2')
efficient_sam_vits_model.device
rle = run_everything_ours(cv2.resize(image,(512,512)), efficient_sam_vits_model)

**Comment by xiaoqiang-lu** [comment]
Me too, OOM occured using v100-32G.

**Comment by yformer** [comment]
@yihong1120 @WanBenLe @xiaoqiang-lu, we released efficient track anything for unified image and video segmentation, which provides a family of models for quality and latency needs, https://github.com/yformer/EfficientTAM
